File: 221-maximal-square/221-maximal-square.java
public class Solution {
    public int maximalSquare(char[][] matrix) {
        int rows = matrix.length, cols = rows>0 ? matrix[0].length : 0;
        int[][] dp = new int[rows+1][cols+1];
        int maxqlen = 0;
        
        for(int i = 1; i <= rows; i++){
            for(int j = 1; j <= cols; j++){
                if(matrix[i-1][j-1] == '1'){
                    dp[i][j] = Math.min(dp[i-1][j],Math.min(dp[i-1][j-1],dp[i][j-1])) + 1;
                    maxqlen = Math.max(dp[i][j],maxqlen);
                }
            }
        }
        return maxqlen * maxqlen;
    }
}

        

File: 221-maximal-square/221-maximal-square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        vector<vector<int>> dp(matrix.size(),vector<int>(matrix[0].size()));
        int ROW = matrix.size();
        int COL = matrix[0].size();
        int area = 0;
        for(int i = 0; i < ROW; i++){
            for(int j = 0; j < COL; j++){
                if(i == 0 or j == 0){
                    dp[i][j] = matrix[i][j] == '1' ? 1 : 0;
                    area = max(area,dp[i][j]);
                }else if(matrix[i][j] == '1'){
                    dp[i][j] = min(min(dp[i-1][j],dp[i][j-1]),dp[i-1][j-1]) + 1;
                    area = max(area,dp[i][j]);
                }
            }
        }
        return area * area;
    }
};